metrics/metrics.go

```diff
@@ -0,0 +1,37 @@
+// Package metrics provides an extensible framework to instrument your
+// application. All metrics are safe for concurrent use. Considerable design
+// influence has been taken from https://github.com/codahale/metrics and
+// https://prometheus.io.
+package metrics
+
+// Counter is a monotonically-increasing, unsigned, 64-bit integer used to
+// capture the number of times an event has occurred. By tracking the deltas
+// between measurements of a counter over intervals of time, an aggregation
+// layer can derive rates, acceleration, etc.
+type Counter interface {
+	With(Field) Counter
+	Add(delta uint64)
+}
+
+// Gauge captures instantaneous measurements of something using signed, 64-bit
+// integers. The value does not need to be monotonic.
+type Gauge interface {
```

Comment:
my experience with Gauges suggests that they are better implemented by callbacks than by Set() sorts of interfaces... I actually think there's something to be said for allowing Counters to be [optionally] implemented that way, but it's less important in that case.